Source/console-checkers/Source/Game.cpp

```diff
@@ -0,0 +1,142 @@
+		std::cend(m_registeredGameBoardViews),
+		[id](const std::unique_ptr<IGameBoardViewStrategy>& view)
+	{
+		return view->GetId() == id;
+	});
+
+	return it != std::cend(m_registeredGameBoardViews);
+}
+
+IGameBoardViewStrategy* GameBoardViewStrategyRegistry::GetGameBoardViewStrategyForId(GameBoardViewStrategyId id) const
+{
+	auto it = std::find_if(std::cbegin(m_registeredGameBoardViews),
+		std::cend(m_registeredGameBoardViews),
+		[id](const std::unique_ptr<IGameBoardViewStrategy>& view)
+	{
+		return view->GetId() == id;
+	});
+
+	if (it != std::cend(m_registeredGameBoardViews))
+	{
+		return it->get();
+	}
+
+#ifdef DEBUG
+	// Hi! If you're here, you forgot to add the view to the registry :).
+	assert(false);
+#endif
+
+	spdlog::error("Attempted to use unregistered game view strategy. id={}", id);
+	return nullptr;
+}
+
+IGameBoardViewStrategy* GameBoardViewStrategyRegistry::GetGameBoardViewStrategyForPlayerOption(int32_t option) const
+{
+	// Options are 1 based because they are player-facing. We subtract one for indexing.
+	option--;
+
+	if (option < 0 || option > static_cast<int32_t>(m_registeredGameBoardViews.size()))
+	{
+		// This can happen if the player selected an option that we deemed to be a valid number, but doesn't
+		// actually map to anything we have registered.
+		return nullptr;
+	}
+	return m_registeredGameBoardViews[option].get();
+}
+
+//---------------------------------------------------------------
+
+Game::Game()
+	: m_inputComponent(std::make_unique<ConsoleInputComponent>(this))
+	, m_gameState(std::make_unique<GameState>(this))
+	, m_gameBoardViewStrategyRegistry(std::make_unique<GameBoardViewStrategyRegistry>())
+{
+	m_selectedGameBoardViewStrategy = m_gameBoardViewStrategyRegistry->GetGameBoardViewStrategyForId(
+		GameplaySettings::s_defaultGameBoardViewStrategy);
+
+	// TODO: Its not obvious that we CANNOT fire off UI events in here. Display
+	// is not alive yet. But we don't know that. Our owner should probably tell us.
+	// Or Display could via UI events but that seems a bit out of place.
+}
+
+Game::~Game() = default;
+
+void Game::SetSelectedGameBoardViewStrategy(GameBoardViewStrategyId id)
+{
+	if (m_selectedGameBoardViewStrategy->GetId() == id)
+	{
+		return;
+	}
+
+	m_selectedGameBoardViewStrategy = m_gameBoardViewStrategyRegistry->GetGameBoardViewStrategyForId(id);
+
+	// Our UI needs to be notified of the change.
+	m_uiPromptRequestedEvents.GetGameBoardViewStrategyChangedEvent().notify(m_selectedGameBoardViewStrategy);
+}
+
+void Game::MovePiece(const PieceMoveDescription& moveDescription) const
+{
+	m_gameState->MovePiece(moveDescription);
+}
+
+void Game::Run()
+{
+	m_uiPromptRequestedEvents.GetGameBoardViewStrategyChangedEvent().notify(m_selectedGameBoardViewStrategy);
+	m_uiPromptRequestedEvents.GetWelcomePromptRequestedEvent().notify();
+
+	m_gameState->ToggleTurnPlayer();
+
+	while (m_gameState->GetWinState() == WinConditionReason::None)
```

Comment:
This doesn't burn at all actually heh. its input bound. Everything is event driven. The game loop is

Wait for input
  OnInput
    handleCommand
    mutateGameState
    NotifyView

Reply:
Ya it's not obvious to the reader that RequestAndProcessInput blocks on input. I would _never_ expect a game's loop to block. It just poorly models how games are architected. Game's continue to process all sorts of stuff without any user input - sound, rendering, AI, networking, etc. Currently while you're waiting for input nothing else can happen.

You need to unblock your main loop and figure out how to handle input and output from the console. You seem convinced this is "impossible" but is it not just a concurrency problem? I would imagine you have a thread which handles input and output from the console. Just like how Phx/Agt comms don't block Phx main thread. And you wouldn't "render" (aka write to console) 60 times a sec. You would once, when the GameState is dirty and needs to be drawn. Console IS a GUI. Just a horrible one with dogshit restrictions lol.

I would offer solving how to architect this program so that console or SFML or any other view can be dropped in with a smallish amount of changes would be a senior II level problem and be highly educational. Same goes for handling input which can be abstracted to not know input is coming from the console instead of hotkeys. Stupid? Ya, console is pretty stupid lol. Hard? God yes. Educational? Enormously.